keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+*Alternative: many flags*
+
+By default, plugins not listed in a trusted catalog will not be executed and an error will be printed. The providers for these plugins must be listed explicitly in each plugin config, and additional flags are always required to use them. Which flags are required depends on the provider type.
+
+For plugins distributed as containers, the `--enable-container-plugins` flag will need to be provided. The containers will continue to be denied disk and network access by default. To enable access, the existing network/mount restriction flags will be retained but renamed to clarify what they govern:  `--plugin-container-networks=[STRING]` `--plugin-container-mounts=[STRING]`.
+
+For plugins distributed as executables (typically committed in git along with the Kustomization), every provider the Kustomization needs must be explicitly named in a new flag:  `--trust-embedded-exec-plugins=base/plugin/transformer.sh,overlay/scripts/gen.sh`. By using this flag, the user demonstrates they are aware of what will be executed and accept the risk of running each one. To make the list less painful to construct, the error message thrown when this list is not provided or is incomplete will inform the user of which are missing.
+
+**Additional constraints for exec providers**
+
+The location of exec providers that are local will be subject Kustomize's load restrictor, i.e. the LoadRestrictionsRootOnly policy will apply to them by default (it does not today). If this is deemed insufficient, we could take it a step further and force the LoadRestrictionsRootOnly policy even when a different loader policy is provided by flag.
+
+### Supported runtimes
+
+Container and exec KRM Function plugins will continue to be supported. 
+
+Starlark support will be deprecated and removed. This is because the dependencies the Starlark runtime brings in were deemed unacceptable for inclusion in kubectl, which would lead to a permanent discrepancy between kustomize distributions (an anti-goal of this KEP). No Starlark SDK has been developed to date, and Starlark scripts can still be used, like any other script, in the context of a container. This is the path kpt has chosen as well with [StarlarkRun](https://kpt.dev/book/05-developing-functions/04-executable-configuration?id=starlark-function).
+
+Go plugins are currently only supported by legacy plugins, not by the KRM Function plugin style. Although these have advantages in terms of execution speed and ease of testing, they have many shortcomings and cannot provide the level of user experience one would expect from a "plugin". Most notably, they are not portable, requiring end users to undertake compilation steps themselves. For these reasons, this KEP does not recommend porting support to KRM Function plugins as part of plugin graduation. Nothing prevents it from being added in the future if there is sufficient interest. 
```

Comment:
+1 here and I am curious what it would take to deprecate those.  One of the benefits of kustomize being a distributed binary is that a user can stick to whatever version they like until they are ready to make a jump to the new architecture.  This is much better/easier than a hosted API.

Reply:
Since they are alpha, I don't think we need much of a deprecation process. That said, the way Kustomize is implemented internally actually leverages these, so it may be much more effort than it is worth to remove them entirely from the codebase.